crypto/evp/digest.c

```diff
@@ -494,14 +498,27 @@ int EVP_Digest(const void *data, size_t count,
     return ret;
 }
 
+int EVP_MD_CTX_setparams(EVP_MD_CTX *ctx, OSSL_PARAM *params)
+{
+    if (ctx->digest != NULL && ctx->digest->setparams != NULL)
+        return ctx->digest->setparams(ctx->provctx, params);
+    return 0;
+}
+
 int EVP_MD_CTX_ctrl(EVP_MD_CTX *ctx, int cmd, int p1, void *p2)
 {
     if (ctx->digest != NULL) {
-        if (ctx->digest->dctrl != NULL) {
-            int ret = ctx->digest->dctrl(ctx->provctx, cmd, p1, p2);
-            if (ret <= 0)
-                return 0;
-            return 1;
+        if (ctx->digest->setparams != NULL) {
+            OSSL_PARAM params[4] = {
+                OSSL_PARAM_int(OSSL_PARAM_DIGEST_KEY_CMD, &cmd),
+                OSSL_PARAM_int(OSSL_PARAM_DIGEST_KEY_P1, &p1),
+                OSSL_PARAM_octet_ptr(OSSL_PARAM_DIGEST_KEY_P2, p2, p1),
+                OSSL_PARAM_END
+            };
```

Comment:
I see where you're going with this, but I think this is flawed.  Basically, with OSSL_PARAM arrays, the ctrl() type of function become legacy, so for *providers*, we can choose to translate only the controls that we know about.  In other words, something like this:

``` C
            OSSL_PARAM params[2];
            size_t n = 0;
            size_t retsize = 0;

            switch (cmd) {
            case EVP_MD_CTRL_DIGALGID:
                params[n++] = OSSL_PARAM_construct_utf8_string(OSSL_PARAM_DIGEST_ALGID, p2, p1 ? p1 : 9999, &retsize);
                params[n++] = OSSL_PARAM_end();
                if (ctx->digest->get_params(ctx->provctx, params))
                    return 1;
                break;
            case EVP_MD_CTRL_XOF_LEN:
                params[n++] = OSSL_PARAM_construct_int(OSSL_PARAM_DIGEST_ALGID, &p1, NULL);
                params[n++] = OSSL_PARAM_end();
                if (ctx->digest->set_params(ctx->provctx, params))
                    return 1;
                break;
            }
            return 0;
```

Do note the fact that one of the controls shown above is a *getter*, while the other one is a *setter*, hence the `get_params` vs `set_params` calls.  That's one of the tricky bits with ctrl() functions, that they can act as both.

So why only support the controls that *we* know?  It's really quite simple, if we put into our minds that providers will use OSSL_PARAM mechanisms instead, then ctrl() functions squarely belong with legacy code, and any control command we do not know about will belong with legacy methods, such as engines.  So we really only need to "translate" the stuff that we know.

Reply:
Both EVP_MD_CTRL_MICALG & EVP_MD_CTRL_DIGALGID look like dead code to me (I cant see any digest ctrl methods that access them). So I dont think we should add any code for them (which could then changes behaviour).

Reply:
```
: ; git grep EVP_MD_CTRL_MICALG
CHANGES:     the ctrl EVP_MD_CTRL_MICALG.
crypto/asn1/asn_mime.c:            rv = md->md_ctrl(NULL, EVP_MD_CTRL_MICALG, 0, &micstr);
doc/man3/EVP_DigestInit.pod:=item EVP_MD_CTRL_MICALG
include/openssl/evp.h:#  define EVP_MD_CTRL_MICALG                      0x2
```

Not entirely dead if we call it...  and looking at gost-engine, we find this:

https://github.com/gost-engine/engine/blob/a7a5ff4a0eaff8586f963d7d15a3b3236201c5d0/gost_md2012.c#L137-L168

The question is what a gost provider would look like, or if there is any initiative in that direction.  @beldmit may be able to give us some answers (as soon as he catches up)...

Reply:
Sure. 
According to RFC 3851, 
"The multipart/signed Content type has two required parameters: the protocol parameter and the micalg parameter."

micalg parameter is a property of the hash algorithm used for signing the message with the specified signature algorithm. Before the GOST support appeared, there was a hard-coded 'case md5/sha1/sha256/etc' structure. 

It seems much more reasonable to avoid hard-coding here and replace it with a callback, especially for engine-provided digests.

Reply:
Added the micalg back into the ctrl.